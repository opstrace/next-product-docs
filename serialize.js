module.exports = require('./dist/serialize')
