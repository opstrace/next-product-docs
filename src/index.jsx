import React from 'react'
import { MDXRemote } from 'next-mdx-remote'
import { State, Observe } from 'mdx-observable'
import { Element } from 'react-scroll'
import CodeBlock from './components/CodeBlock.jsx'
import InlineCode from './components/InlineCode.jsx'
import Tabs, { Tab } from './components/Tabs.jsx'
import Interpolate from './components/Interpolate.jsx'

export default function Documentation({ source, theme }) {
  const components = {
    Tabs: Tabs,
    Tab: Tab,
    Interpolate,
    State,
    Observe,
    // eslint-disable-next-line react/display-name
    Element: ({ name, ...props }) => {
      return (
        <Element
          // remove name from parent div
          name={props.children[0]?.props?.id === name ? null : name}
          {...props}
        />
      )
    },
    // eslint-disable-next-line react/display-name
    pre: (props) => <div {...props} />,
    code: CodeBlock,
    inlineCode: InlineCode
  }

  return <MDXRemote {...source} components={components} />
}
