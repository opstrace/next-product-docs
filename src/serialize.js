import { serialize } from 'next-mdx-remote/serialize'
import matter from 'gray-matter'
import { marked } from 'marked'
import GithubSlugger from 'github-slugger'
import rehypeSlug from 'rehype-slug'
import rehypeAutolink from 'rehype-autolink-headings'
import cloneDeep from 'lodash/cloneDeep'

import remarkTabs from './lib/remark-plugins/tabs'
import remarkState from './lib/remark-plugins/state'
import remarkSections from './lib/remark-plugins/sections'
import remarkExternalLinks from 'remark-external-links'
import remarkInternalLinks from './lib/remark-plugins/links'
import remarkRewriteImages from './lib/remark-plugins/images'
import {
  findRouteByPath,
  replaceDefaultPath,
  getSlug,
  getPaths,
  fetchDocsManifest
} from './lib/docs'
import { getRawFile } from './lib/files'

const DOCS_FOLDER = process.env.DOCS_FOLDER

export async function pageProps({ params }) {
  const slugger = new GithubSlugger()
  const manifest = await fetchDocsManifest().catch((error) => {
    if (error.status === 404) return
    throw error
  })
  const { slug } = getSlug(params)
  const route = manifest && findRouteByPath(slug, manifest.routes)
  if (!route)
    return {
      notFound: true
    }
  const manifestRoutes = cloneDeep(manifest.routes)
  replaceDefaultPath(manifestRoutes)

  const mdxRawContent = await getRawFile(route.path)
  const { content, data } = matter(mdxRawContent)
  if (!data.title) {
    data.title = ''
  }
  const mdxSource = await serialize(content, {
    scope: { data },
    mdxOptions: {
      rehypePlugins: [
        rehypeSlug,
        [
          rehypeAutolink,
          {
            behavior: 'append',
            properties: {
              className: ['btn-copy-link', 'invisible']
            },
            content: {
              type: 'element',
              tagName: 'svg',
              properties: {
                className: ['h-6', 'w-6', 'ml-2', 'docs-copy-btn'],
                xmlns: 'http://www.w3.org/2000/svg',
                fill: 'none',
                viewBox: '0 0 24 24',
                stroke: 'currentColor'
              },
              children: [
                {
                  type: 'element',
                  tagName: 'path',
                  properties: {
                    strokeLinecap: 'round',
                    strokeLinejoin: 'round',
                    strokeWidth: 2,
                    d: 'M13.828 10.172a4 4 0 00-5.656 0l-4 4a4 4 0 105.656 5.656l1.102-1.101m-.758-4.899a4 4 0 005.656 0l4-4a4 4 0 00-5.656-5.656l-1.1 1.1'
                  },
                  children: []
                }
              ]
            }
          }
        ]
      ],
      remarkPlugins: [
        remarkSections,
        remarkTabs,
        remarkState,
        [remarkRewriteImages, { destination: process.env.ASSETS_DESTINATION }],
        [remarkExternalLinks, { target: false, rel: ['nofollow'] }],
        [
          remarkInternalLinks,
          {
            prefix: DOCS_FOLDER,
            slug: params.slug,
            extensions: ['.mdx', '.md']
          }
        ]
      ]
    },
    target: ['es2020']
  })

  const markdownTokens = marked.lexer(content)
  const headings = markdownTokens
    .filter((t) => t.type === 'heading')
    .map((heading) => {
      heading.slug = slugger.slug(heading.text)
      return heading
    })

  const title = headings && headings.length > 0 ? headings[0].text : data.title

  return {
    title,
    frontmatter: data,
    sidebarRoutes: manifestRoutes,
    route,
    source: mdxSource,
    tocHeadings: headings
  }
}

export async function staticPaths() {
  const manifest = await fetchDocsManifest()
  const paths = getPaths(manifest.routes)
  paths.shift() // remove "/docs/README"
  paths.unshift(`/${process.env.DOCS_FOLDER}`)
  return paths
}
