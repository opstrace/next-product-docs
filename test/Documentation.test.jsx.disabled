import { render } from '@testing-library/react'
import { Documentation } from '../src/'
import { pageProps } from '../src/serialize'
import nock from 'nock'
import { describe, it, expect, beforeEach } from 'vitest'
import 'whatwg-fetch'
import { afterAll } from 'vitest'

const { ...originalEnv } = process.env

beforeEach(async () => {
  const docOneFixture = `
  ---
  description: Some description.
  ---

  # First Level

  ## Second Level One

  First Paragraph.

  ## Second Level Two

  Second Paragraph.
  `

  process.env.DOCS_ORG = 'DOCS_ORG'
  process.env.DOCS_REPO = 'DOCS_REPO'
  process.env.DOCS_BRANCH = 'DOCS_BRANCH'
  process.env.DOCS_FOLDER = 'DOCS_FOLDER'

  // mock github response
  nock(
    `https://raw.githubusercontent.com/${process.env.DOCS_ORG}/${process.env.DOCS_REPO}/${process.env.DOCS_BRANCH}`
  )
    .defaultReplyHeaders({
      'access-control-allow-origin': '*',
      'access-control-allow-credentials': 'true'
    })
    .get(`/${process.env.DOCS_FOLDER}/manifest.json`)
    .reply(200, {
      routes: [
        {
          heading: true,
          title: 'Top level',
          routes: [
            {
              title: 'Docs One',
              path: `/${process.env.DOCS_FOLDER}/one.md`
            }
          ]
        }
      ]
    })
    .get(`/${process.env.DOCS_FOLDER}/one.md`)
    .reply(200, docOneFixture)
    .persist()
})

afterAll(async () => {
  process.env = originalEnv
})

describe('Document', () => {
  it('should render', async () => {
    const { source } = await pageProps({ params: { slug: ['one'] } })
    const { container } = render(<Documentation source={source} />)
    expect(container).toMatchSnapshot()
  })
})
