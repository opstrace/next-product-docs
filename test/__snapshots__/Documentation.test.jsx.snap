// Vitest Snapshot v1

exports[`Document > should render 1`] = `
<div>
  &lt;State initialstate={
      {
        
      }
    }&gt;
  

  {({setState: setSubstitutionState, ...substitutionState}) =&gt; (
  

  &lt;&gt;
  

  &lt;Interpolate substitutions={substitutionState}&gt;
  

  <hr />
  

  <h2
    id="description-some-description"
  >
    description: Some description.
    <a
      class="btn-copy-link invisible"
      href="#description-some-description"
    >
      <svg
        class="h-6 w-6 ml-2 docs-copy-btn"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M13.828 10.172a4 4 0 00-5.656 0l-4 4a4 4 0 105.656 5.656l1.102-1.101m-.758-4.899a4 4 0 005.656 0l4-4a4 4 0 00-5.656-5.656l-1.1 1.1"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </a>
  </h2>
  

  <h1
    id="first-level"
  >
    First Level
    <a
      class="btn-copy-link invisible"
      href="#first-level"
    >
      <svg
        class="h-6 w-6 ml-2 docs-copy-btn"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M13.828 10.172a4 4 0 00-5.656 0l-4 4a4 4 0 105.656 5.656l1.102-1.101m-.758-4.899a4 4 0 005.656 0l4-4a4 4 0 00-5.656-5.656l-1.1 1.1"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </a>
  </h1>
  

  <h2
    id="second-level-one"
  >
    Second Level One
    <a
      class="btn-copy-link invisible"
      href="#second-level-one"
    >
      <svg
        class="h-6 w-6 ml-2 docs-copy-btn"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M13.828 10.172a4 4 0 00-5.656 0l-4 4a4 4 0 105.656 5.656l1.102-1.101m-.758-4.899a4 4 0 005.656 0l4-4a4 4 0 00-5.656-5.656l-1.1 1.1"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </a>
  </h2>
  

  <p>
    First Paragraph.
  </p>
  

  <h2
    id="second-level-two"
  >
    Second Level Two
    <a
      class="btn-copy-link invisible"
      href="#second-level-two"
    >
      <svg
        class="h-6 w-6 ml-2 docs-copy-btn"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M13.828 10.172a4 4 0 00-5.656 0l-4 4a4 4 0 105.656 5.656l1.102-1.101m-.758-4.899a4 4 0 005.656 0l4-4a4 4 0 00-5.656-5.656l-1.1 1.1"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </a>
  </h2>
  

  <p>
    Second Paragraph.
  </p>
  

  &lt;/Interpolate&gt;
  

  &lt;/&gt;
  

  )}
  

  &lt;/State&gt;
</div>
`;
